subsys/bluetooth/mesh/gatt_cli.c

```diff
@@ -0,0 +1,374 @@
+/*
+ * Copyright (c) 2021 Xiaomi Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <sys/byteorder.h>
+
+#include <net/buf.h>
+#include <bluetooth/bluetooth.h>
+#include <bluetooth/hci.h>
+#include <bluetooth/uuid.h>
+#include <bluetooth/conn.h>
+#include <bluetooth/gatt.h>
+#include <bluetooth/mesh.h>
+
+#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_MESH_DEBUG_PROXY)
+#define LOG_MODULE_NAME bt_mesh_gatt_client
+#include "common/log.h"
+
+#include "mesh.h"
+#include "adv.h"
+#include "net.h"
+#include "rpl.h"
+#include "transport.h"
+#include "host/ecc.h"
+#include "prov.h"
+#include "beacon.h"
+#include "foundation.h"
+#include "access.h"
+#include "proxy.h"
+#include "proxy_msg.h"
+#include "proxy_cli.h"
+#include "gatt_cli.h"
+#include "pb_gatt_cli.h"
+
+static struct bt_mesh_gatt_server {
+	struct bt_conn *conn;
+	uint16_t svc_start_handle;
+	uint16_t data_in_handle;
+	const struct bt_mesh_gatt_cli *gatt;
+
+	union {
+		void *user_data;
+		struct bt_gatt_discover_params discover;
+		struct bt_gatt_subscribe_params subscribe;
+	};
+} servers[CONFIG_BT_MAX_CONN];
+
+static struct bt_mesh_gatt_server *find_server(struct bt_conn *conn)
+{
+	for (int i = 0; i < ARRAY_SIZE(servers); i++) {
+		if (servers[i].conn == conn) {
+			return &servers[i];
+		}
+	}
+
+	return NULL;
+}
+
+static uint8_t notify_func(struct bt_conn *conn,
+			   struct bt_gatt_subscribe_params *params,
+			   const void *data, uint16_t length)
+{
+	const uint8_t *val = data;
+	struct bt_mesh_gatt_server *server;
+
+	if (!data) {
+		BT_WARN("[UNSUBSCRIBED]");
+		params->value_handle = 0U;
+		return BT_GATT_ITER_STOP;
+	}
+
+	server = find_server(conn);
+	if (!server) {
+		return BT_GATT_ITER_STOP;
+	}
```

Comment:
Once you start using `bt_conn_index()` you'll never fail with allocating contexts, so this branch becomes dead code